client/client.js

```diff
@@ -78,67 +79,85 @@ new Vue({
       })
     },
     initialiseStatshistory: function(){
-      this.statshistory = [];
-      for (var i = 0; i < numXValues; i++) {
-      	this.statshistory.push(
-      	  {
-      	    Queue: 0,
-      	    Running: 0,
-      	    Complete: 0,
-      	    FunctionStatsMap: {}
-      	  }      	  
-      	)
-      } 
+      if (this.statshistory==null){
+        this.statshistory = [];
+        for (var i = 0; i < numXValues; i++) {
+          this.statshistory.push(
+            {
+              Queue: 0,
+              Running: 0,
+              Complete: 0,
+              FunctionStatsMap: {}
+            }      	  
+          )
+        } 
+      }
     },    
     loadStats: function(){
       var t = this;
-      $.ajax({
-        url: '/api/stats',
-        dataType: 'json',
-        success: (statistics) => {
-          t.stats = statistics;
-          if (t.statshistory==null){
-            t.statshistory = [statistics];
-          } else {
-            t.statshistory.push(statistics);
-            if (t.statshistory.length > numXValues){
-              t.statshistory.shift();
-            }
-          }        
-          
-          // do the stats contain a new function?
-          var previousKnownFunctions = Object.keys(t.statshistory[t.statshistory.length-2].FunctionStatsMap);
-          var nowKnownFunctions = Object.keys(t.statshistory[t.statshistory.length-1].FunctionStatsMap);
-          for (var j = 0; j < nowKnownFunctions.length; j++){
-            if (previousKnownFunctions.indexOf(nowKnownFunctions[j])==-1){
-              var newFunction = nowKnownFunctions[j];
-              // we have a new function: backfill all the earlier stats with zero values for this function
-              for (var k = 0; k < t.statshistory.length-1; k++){
-                t.statshistory[k].FunctionStatsMap[newFunction]={Queue:0, Running:0, Complete:0, Failed:0};
+      if (this.autorefresh) {
+        $.ajax({
+          url: '/api/stats',
+          dataType: 'json',
+          success: (statistics) => {
```

Comment:
There's a lot going on in this callback to the point where I'd recommend declaring the success function elsewhere (with a suitably descriptive name) which gets passed to success. This will make it easier for the reader to scan what the intent of the call is and what happens on success/failure without having to go through how it's doing it.

Reply:
Good idea. I'll move it to a named function.